server/vcs/provider/github/list_iterator.go

```diff
@@ -0,0 +1,35 @@
+package github
+
+import (
+	"context"
+	"fmt"
+	gh "github.com/google/go-github/v45/github"
+	"github.com/pkg/errors"
+	"net/http"
+)
+
+// Add supported types here with each GH api fxn introduced
+
+func Iterate[T []*gh.CommitFile | []*gh.PullRequestReview | []*gh.CheckRun, R []string](
```

Comment:
The point of generics is to either use the type of a common interface (in this case there is nothing common so i just use interface{} but that's ok because it still provides compile time typing so you don't ever have to cast)

Also it doesn't make sense to use arrays as the types here.  `R []string` means that the type is effectively always an array of type strings, so at that point why even use a generic type).

This is how I'd write it:

```go
func Iterate[T interface{}](
	ctx context.Context,
	runFunc func(ctx context.Context, nextPage int) ([]T, *gh.Response, error)) ([]T, error) {
	var output []T
	nextPage := 0
	for {
		results, resp, err := runFunc(ctx, nextPage)
		if err != nil {
			return nil, errors.Wrap(err, "error running gh api call")
		}
		if resp.StatusCode != http.StatusOK {
			return nil, fmt.Errorf("not ok status running gh api call: %s", resp.Status)
		}
		output = append(output, results...)
		if resp.NextPage == 0 {
			break
		}
		nextPage = resp.NextPage
	}
	return output, nil
}
```

And then when you want to use it you could do the following:

```go
...

run := func(ctx context.Context, nextPage int) ([]*gh.CommitFile, *gh.Response, error) {
  listOptions := gh.ListOptions{
	  PerPage: 100,
  }
  listOptions.Page = nextPage
  repositoryCommit, resp, err := client.Repositories.GetCommit(ctx, repo.Owner, repo.Name, fileFetcherOptions.Sha, &listOptions)
  if repositoryCommit != nil {
	  return repositoryCommit.Files, resp, err
  }
  return nil, nil, errors.New("unable to retrieve commit files from GH commit")
}

files, err := Iterate(ctx, run)

if err != nil {
   // do something
}

var renamed []string
for _, f := range files {
  renamed = append(renamed, f.GetFilename())
  
  // If the file was renamed, we'll want to run plan in the directory
  // it was moved from as well.
  if f.GetStatus() == "renamed" {
	  renamed = append(renamed, f.GetPreviousFilename())
  }
}

return renamed, nil
```